Replace the `Edges` enumeration values:

desruisseaux opened the pull request:
Replace the `Edges` enumeration values:
    
* PLANAR → LINEAR
* SPHERICAL → PSEUDO_GEODESIC

The reason is that `PLANAR` does not said what the interpolation is. A quadratic or cubic interpolation can also be on the plane. Likewise, the edges on a sphere could be [rhumb lines](https://en.wikipedia.org/wiki/Rhumb_line) instead of geodesics.

Furthermore, the use of `LINEAR` term is consistent with existing practices in the OGC standards that I'm aware of. See for example [OGC Moving Features](https://docs.ogc.org/bp/16-140r1/16-140r1.html#simple) > _§6.3.1. Simple Temporal Geometries_ > _Predefined Interpolation Methods_. By contrast, I'm not aware of any OGC specification that would use "planar" for what `Edges` wants to said.

The "pseudo" in `PSEUDO_GEODESIC` is by analogy with the Pseudo-Mercator projection (a.k.a. "Google Mercator" or "Popular visualization"). This is exactly the same situation as Pseudo-Mercator: applying spherical formulas on an ellipsoid, ignoring the flattening. The explanatory text has been modified since it was not quite right. In particular, if the specification wants to be unambiguous, it needs to said what to do with the latitude values even if that action is "do nothing" (more advanced applications would have converted geodetic latitudes to authalic latitudes before to apply the spherical formulas. See for example ESRI's _"Mercator Auxiliary Sphere"_ projection).

wgtmac committed:
WIP: Add geometry logical type

wgtmac committed:
address various comments

wgtmac committed:
add file level geo stats

wgtmac committed:
address feedback:
- remove file-level geo stats
- add custom wkb-encoded geometry stats
- comment out controversial items

wgtmac committed:
change naming and remove controversial items

wgtmac committed:
address feedback

wgtmac committed:
fix typo

wgtmac committed:
use WKB type code

zhangfengcdt committed:
Update covering and geometry type protocol based on comments (#2)

jiayuasu committed:
Add the new suggestion according to the meeting with Snowflake (#3)

wgtmac committed:
change metadata to string type and rewording WKB description

wgtmac committed:
add example for crs

wgtmac committed:
reword crs

wgtmac committed:
clarify WKB

wgtmac committed:
clarify coverings

jiayuasu committed:
Update the suggestion for bbox stats (#4)

* Add the new suggestion according to the meeting with Snowflake

* Refine the description according to the suggestion

wgtmac committed:
Update src/main/thrift/parquet.thrift

Co-authored-by: Dewey Dunnington <dewey@dunnington.ca>

wgtmac committed:
Update src/main/thrift/parquet.thrift

Co-authored-by: Dewey Dunnington <dewey@dunnington.ca>

wgtmac committed:
Update src/main/thrift/parquet.thrift

Co-authored-by: Dewey Dunnington <dewey@dunnington.ca>

wgtmac committed:
Update src/main/thrift/parquet.thrift

Co-authored-by: Dewey Dunnington <dewey@dunnington.ca>

wgtmac committed:
address feedback about edges and wkb

wgtmac committed:
add geoparquet column metadata back

jiayuasu committed:
Update the spec according to the new feedback (#5)

* Update the spec according to the new feedback

* Fix typo

desruisseaux committed:
Replace the `Edges` enumeration values and update the `BoundingBox` documentation accordingly.

* PLANAR → LINEAR
* SPHERICAL → PSEUDO_GEODESIC

Rational: https://github.com/wgtmac/parquet-format/pull/6

jorisvandenbossche commented on src/main/thrift/parquet.thrift:
```diff
@@ -269,36 +269,43 @@ enum GeometryEncoding {
- * Because most systems currently assume planar edges and do not support
- * spherical edges, planar should be used as the default value.
+ * The LINEAR value is independent of the coordinate reference system.
+ * The PSEUDO_GEODESIC value requires a coordinate reference system associated
```
I think the "pseudo geodesic" is also fine, but for reference, S2 calls this a "spherical geodesic":

From http://s2geometry.io/about/overview (in "Design Choices", can't link to the exact section):

> In S2, all edges are “spherical geodesics”, i.e. shortest paths on the sphere

desruisseaux replied:
I wasn't aware of that, thank for the link. I just checked whether their use of "spherical geodesic" is in the same sense as the proposed "pseudo geodesic". I believe that there is a nuance between S2 usage and this PR. The above-cited S2 overview also said:

> Geocentric and geodetic coordinates are two methods for defining a (latitude, longitude) coordinate system over the Earth’s surface. The question sometimes arises: which system does the S2 library use? The answer is: neither (or both). Points in S2 are modeled as lying on a perfect sphere

Geocentric and geodetic latitudes are the same on a sphere, but differ on an ellipsoid. So saying that S2 uses none or both of them is indeed equivalent to saying that S2 works only with coordinates on a sphere. In other words, if S2 was defining a CRS (I saw none in the overview), it would be required to be associated to a sphere, not to an ellipsoid. In this context, "spherical geodesic" is a true geodesic, with the word "spherical" only emphasizing that they are working on a sphere all the way down to the very definition of what is a latitude. Next to that, the S2 overview also said:

> How points are mapped onto the sphere is outside the scope of S2

In my understanding, the Parquet geometry extension is where this mapping would be defined, and it seems to be implicitly a "do nothing" operation. More advanced applications would, for example, convert geodetic latitudes to some [auxiliary latitudes](https://mathworld.wolfram.com/AuxiliaryLatitude.html) (e.g. authalic latitudes or conformal latitudes) before to give them to S2, but this is another story. The key point is that the decision to approximate an ellipsoid by a sphere (and how to do that) is done in Parquet, not in S2. So the fact that geodesics are not real geodesics was not really relevant to S2, but become relevant to Parquet.

desruisseaux replied:
The problem with "spherical" is similar to `PLANAR`: it does not distinguish between variants of the "spherical geodesic" interpolation method. A future Parquet version could add `AUTHALIC_GEODESIC` (I don't know if it is okay to put those terms together, we would need to check usage) for meaning _"convert geodetic latitudes to authalic latitudes before to compute geodesics on the authalic sphere"_, or `CONFORMAL_GEODESIC` for meaning _"convert to conformal latitudes etc."_. The problem with `SPHERICAL_GEODESIC` is that it applies to all of them. This is similar to `PLANAR` which applies to linear interpolation, quadratic, cubic, Bézier, _etc._

**Note 1:** users would be interested in the authalic variant if they care about surfaces, or in the conformal variant if they care about shapes. But I don't know if such usage exists, as a better alternative is to use real geodesics. However, such usage exists for the Mercator projection (ESRI's _"Mercator Auxiliary Sphere type 3"_), which make me think that it could exist for geodesics too.

**Note 2:** it is quite possible that the group decides that Parquet will never add such beasts as "authalic geodesic", and instead add real geodesics if demanded. In such case, `SPHERICAL_GEODESIC` is as good as other names. But it would be a constraint on future maintainers. `PSEUDO_GEODESIC` seems more future-proof to me, as it leaves more doors open.

desruisseaux replied:
I'm not sure to understand. Addition of `Edges` enum values as the right door for more accurate representations in the future was precisely the reason why I suggested to avoid "spherical", as it would clash (i.e., create a confusing situation) with the new enum values…

jiayuasu replied:
@desruisseaux @paleolimbot Do we achieve any agreement on this?

jiayuasu commented:
CC @szehon-ho for the Iceberg geo spec

desruisseaux commented on src/main/thrift/parquet.thrift:
```diff
@@ -269,36 +269,43 @@ enum GeometryEncoding {
  */
 struct BoundingBox {
-  /** Westmost value if edges = spherical **/
+  /** Westmost value if the coordinate system is spherical **/
```
No problem, but then the word to change would be `spherical`. Maybe "if edges are geodesic"? (without "spherical" in order to mean "any kind of geodesic")?

desruisseaux replied:
Changed _"if the coordinate system is spherical"_ by _"if edges are geodesics"_. I propose to use only "geodesics" instead of "pseudo geodesics" or "spherical geodesics" because that section is about min/max being interpreted as westmost/eastmost, which applies to any kind of geodesics. I can rename again if necessary.

desruisseaux committed:
Replace "if the coordinate system is spherical" by "if edges are geodesics".
https://github.com/wgtmac/parquet-format/pull/6#discussion_r1796385072

Note: this commit also contains spurious changes in some other lines
that are only the removal of trailing whitespaces.
